### ISSUE TITLE
Portability?

### ISSUE THREAD

**Issue by jlmcmchl** [argmin-rs/argmin-core#1]
Hi,

I'm in the process of developing a web tool that draws paths for a robot to follow, and it uses a WASM library to turn waypoints into a path. In addition, I have plans for a means to optimize the path for change in curvature, i.e. how smooth the path is (and thus, easy to follow).

Instead of rolling my own gradient descent, I'd like to use argmin. I'm currently unable to do so, however, because the ctrlc dependency doesn't compile for wasm32.

How could I turn ctrlc into a feature instead of a hard dependency? I'm more than happy to do the work required, I'm just lost on how I would begin.

Thanks!

**Comment by stefan-k** [comment]
Hi, 

thanks for giving argmin a try! 
I added the `disable_ctrlc` branch (to `argmin-core`, `argmin` and `argmin-codegen`) which turns `ctrlc` into an opt-in feature. I would prefer it if it could be opt-out but that does not seem to be possible at the moment. It seems to work with the steepestdescent example. It would be great if you could test it and let me know how it works for you.
The `Cargo.toml`s of all projects currently reflect my dev setup, so have a look there if you run into problems with compilation.

I'm actually (pleasantly) surprised that only ctrlc is causing problems when compiling to wasm. Let me know if more problems come up. I haven't considered features so far, but I'm sure there are more things that would make sense to be optional (for instance ndarray support).

**Comment by jlmcmchl** [comment]
Thanks for making the branch! I wish I could say that it fixed everything, but I think I'm closer to 

I went ahead and tried building with the following setup:
toolchain: nightly 10/24
target: wasm32-unknown-unknown
compilation command:
`cargo build --release --target wasm32-unknown-unknown --no-default-features -Z avoid-dev-deps`

It seems like ndarray-linalg is still going to cause a problem, because it relies on lapack and cblas, which won't compile because libc doesn't expose c_double and the like for wasm32. `-Z avoid-dev-deps`, in addition to turning it into an optional dependency, allows me to compile without it.

At this point, I'm getting a fairly cryptic error. I can't imagine why rustc expects something about ctrlc anymore, but it does, about 10 times:
```
error[E0432]: unresolved import `ctrlc`
  --> src/solver/conjugategradient/conjugategradient.rs:24:10
   |
24 | #[derive(ArgminSolver)]
   |          ^^^^^^^^^^^^ no `ctrlc` in the root
```

Do you have any idea why it would be expected?

At this point I'm not convinced that this can be fixed, and that building for wasm32 is going to be more of a curiosity than a reality.

**Comment by stefan-k** [comment]
Did  you also checkout the `disable_ctrlc` branch in `argmin_codegen`? Looks  to me like it is still generating the `ctrlc` code, which it shouldn't with the correct branch. 

Regarding ndarray-linalg (and ndarray): I was thinking about turning everything ndarray related into a feature anyway. The related code is only in `src/math.rs` (which is currently a mess) and mostly created by macros, so it shouldn't be too much work. I'll give it a try and keep you posted.

**Comment by stefan-k** [comment]
`ndarray` and `ndarray-linalg` are now optional features and by default disabled.

I have no experience with wasm. Thanks for posting the compilation command, this has helped me a lot. With this command I can compile `argmin` and `argmin-core`. I could however not figure out how to pass `-Z avoid-dev-deps` to `cargo web`, therefore I can only run the steepest descent example if I comment the dev dependencies in `Cargo.toml` of argmin. This should not be a problem for you if you use argmin as a library, because rustc will ignore the dev dependencies (I think).

I used the following command:

`cargo web start --example steepestdescent --target wasm32-unknown-unknown`

This works, but I'm not sure what I'm supposed to see in the browser. All I see is a blank page ;)

**Comment by stefan-k** [comment]
I published a new version (0.1.2) on crates.io. The features related to `ctrlc`, `ndarray` and `ndarray-linalg` are now disabled by default.

**Comment by jlmcmchl** [comment]
I forgot to mention it here, but my issue earlier was that I didn't update my local copy of argmin-codegen. I've been able to add argmin to my project and build for wasm32-unknown-unknown since you added the feature toggles.

Thanks so much! Closing since this is working now.
